### PR TITLE
Fix: break file-download for prevent ErrnoException

### PR DESCRIPTION
**Pull request by alvince**
Prevent android.system.ErrnoException: write failed: ENOSPC (No space left on device) from OutputResourceWrapper.write()

### CODE REVIEW

File: fetch2/src/main/java/com/tonyodev/fetch2/HttpUrlConnectionDownloader.kt [hunk 1]
```diff
@@ -5,14 +5,15 @@ import java.io.InputStream
 import java.net.CookieHandler
 import java.net.HttpURLConnection
 import java.net.URL
-import java.util.Collections
+import java.util.*
```
Comment (AlexStarc) [inline fetch2/src/main/java/com/tonyodev/fetch2/HttpUrlConnectionDownloader.kt]:
```suggestion
import java.util.Collections
```

File: fetch2/src/main/java/com/tonyodev/fetch2/downloader/FileDownloader.kt [hunk 2]
```diff
@@ -33,4 +35,35 @@ interface FileDownloader : Runnable {
...
+                val spaceLeft = when {
+                    it.startsWith("/data/data") -> {
+                        Environment.getDataDirectory().freeSpace
+                    }
+                    Environment.getExternalStorageState() == Environment.MEDIA_MOUNTED -> {
+                        Environment.getExternalStorageDirectory().freeSpace
+                    }
+                    else -> 0L
+                }
+                spaceLeft > len.times(1.5)
```
Comment (AlexStarc) [inline fetch2/src/main/java/com/tonyodev/fetch2/downloader/FileDownloader.kt]: Why 1.5?

File: fetch2/src/main/java/com/tonyodev/fetch2/downloader/FileDownloader.kt [hunk 3]
```diff
@@ -33,4 +35,35 @@ interface FileDownloader : Runnable {
 
     }
 
-}
\ No newline at end of file
+}
+
+abstract class AbsFileDownloader(
```
Comment (AlexStarc) [inline fetch2/src/main/java/com/tonyodev/fetch2/downloader/FileDownloader.kt:40]: Is this class needed?

**Review by AlexStarc — COMMENTED** [review]
Probably don't need to change formatting to have less differences

File: fetch2/src/main/java/com/tonyodev/fetch2/downloader/FileDownloader.kt (hunk 2 above)
Comment (alvince) [reply fetch2/src/main/java/com/tonyodev/fetch2/downloader/FileDownloader.kt]: For safe-space left only

File: fetch2/src/main/java/com/tonyodev/fetch2/HttpUrlConnectionDownloader.kt (hunk 1 above)
Comment (alvince) [reply fetch2/src/main/java/com/tonyodev/fetch2/HttpUrlConnectionDownloader.kt:8]: yes, formatting is not necessary.
but I think we should apply the newest official stable kotlin code-style for code which modified.

File: fetch2/src/main/java/com/tonyodev/fetch2/downloader/FileDownloader.kt (hunk 3 above)
Comment (alvince) [reply fetch2/src/main/java/com/tonyodev/fetch2/downloader/FileDownloader.kt:40]: Maybe, just added it for wrap shared checking for file-downloader which I think it is consistent

**Comment by alvince** [comment]
> Probably don't need to change formatting to have less differences

yes, formatting is not necessary.
but I think we should apply the newest official stable kotlin code-style for code which modified.

File: fetch2/src/main/java/com/tonyodev/fetch2/downloader/FileDownloader.kt (hunk 2 above)
Comment (AlexStarc) [reply fetch2/src/main/java/com/tonyodev/fetch2/downloader/FileDownloader.kt]: Not sure if it'll really work well for big files

**Comment by tonyofrancis** [comment]
@AlexStarc @alvince does `FetchConfiguration.preAllocateFileOnCreation(true) not work` for this issue? When a file is enqueued and fetch begins processing the request, we pre-allocate the space needed for the download by creating a temp file for the file size. If there is not enough space we throw error: `FILE_ALLOCATION_ERROR`

**Comment by alvince** [comment]
> @AlexStarc @alvince does `FetchConfiguration.preAllocateFileOnCreation(true) not work` for this issue? When a file is enqueued and fetch begins processing the request, we pre-allocate the space needed for the download by creating a temp file for the file size. If there is not enough space we throw error: `FILE_ALLOCATION_ERROR`

@tonyofrancis the option `preAllocateFileOnCreation` by default `DEFAULT_PREALLOCATE_FILE_ON_CREATE` is `true`, there seem to be some problems.
I will check it next.

**Comment by alvince** [comment]
brrrr.
wrong merge target :)
new pr #493

**Pull request closed without merging**